### ISSUE TITLE
Build broken due to older version of protoc (protobuf)

### ISSUE THREAD

**Issue by maehw** [xtech/self-o-mat#21]
Hi,

I've tried to build self-o-mat on a Raspberry Pi with `2021-10-30-raspios-bullseye-armhf-full.img`. I've updated the system and downloaded the dependencies as described in [README.md#clone](https://github.com/xtech/self-o-mat#clone).

The error messages start with the following snippet:

```
[ 48%] Building CXX object CMakeFiles/self_o_mat.app.dir/src/main.cpp.o
In file included from /home/pi/self-o-mat/src/api/BoothApi.h:8,
                 from /home/pi/self-o-mat/src/main.cpp:8:
/home/pi/self-o-mat/src/api/../protobuf/api.pb.h:17:2: error: #error This file was generated by an older version of protoc which is
   17 | #error This file was generated by an older version of protoc which is
      |  ^~~~~
/home/pi/self-o-mat/src/api/../protobuf/api.pb.h:18:2: error: #error incompatible with your Protocol Buffer headers. Please
   18 | #error incompatible with your Protocol Buffer headers.  Please
      |  ^~~~~
/home/pi/self-o-mat/src/api/../protobuf/api.pb.h:19:2: error: #error regenerate this file with a newer version of protoc.
   19 | #error regenerate this file with a newer version of protoc.
      |  ^~~~~
...
```

Any suggestions? Downgrade protobuf (how)? Update the `api.pb.h` file?

Cheers

Edit: `$ less /home/pi/self-o-mat/src/api/../protobuf/api.pb.h`

```
#if GOOGLE_PROTOBUF_VERSION < 3006001
#error This file was generated by a newer version of protoc which is
#error incompatible with your Protocol Buffer headers.  Please update
#error your headers.
#endif
#if 3006001 < GOOGLE_PROTOBUF_MIN_PROTOC_VERSION
#error This file was generated by an older version of protoc which is
#error incompatible with your Protocol Buffer headers.  Please
#error regenerate this file with a newer version of protoc.
#endif
```

**Comment by ClemensElflein** [comment]
I'm not sure why this is happening on your setup. The /src/protobuf/api.pb.h file is not in this repository. Therefore it is generated by your protoc version and should work. Have you maybe copied the repository from another PC?

**Comment by maehw** [comment]
Yes, I've copied the cloned repository from my Linux machine to the Raspberry Pi using `rsync` (omitting the `.git` and `build` directories).

This was in the wiki, but I deleted it:
https://github.com/xtech/self-o-mat/wiki/Configuration/67d6c4adfc5cdf54012bebf0342ca2331d8ef54e

I've not been checking the `Makefile`. But according to your comment it seems that during the build process there are also files generated in the `./src` subdirectory which should not be copied - or at least cleaned before a new build - when not directly cloning the git repo and building it on the Pi.

**Comment by ClemensElflein** [comment]
Correct, you need to start the build on the Raspberry Pi from scratch. I have tried to do cross compilation but failed.

We then had a dedicated Raspberry Pi 4 building and packing the final binaries for our customers.

**Comment by maehw** [comment]
I see. I was looking for a quick way to test my local changes directly on the Pi without having to push & pull - probably even on a public repo. I think `rsync` should do the trick but either using `make clean` after having copied the files or omitting `src/protobuf` should to the trick here.

Side question: have you noticed remarkable speed-up from Raspberry Pi 3 to 4? I am currently working with a Pi 3 and the live stream updates at a rate of 1-3 fps which doesn't really look very fluent. Just curious if using a Pi 4 would improve things here.

**Comment by ClemensElflein** [comment]
The performance question is an interesting one.
The original self-o-mat software was written for Raspberry Pi 3 B+. I optimized it in such a way that it worked on the GPU directly, no X server or window manager needed. The system image was based on the Raspbian lite image (i.e. console only). This worked wonderfully.

But then the Pi 4 came out. And customers told me they'd only buy the box if I upgrade to Pi4 for the additional performance. The problem is that the direct hardware access did not work anymore and I had to install a desktop environment. In the end, the performance dropped a little compared to the Pi 3B+ but people were buying the photo booth again..

Long story short, you either need to use [SFML-PI](https://github.com/mickelson/sfml-pi) or upgrade to a Pi 4 for better performance.. But since Raspberry Pi 3s won't be around forever, I'd vote for upgrading to a Pi 4. I'm not sure about the exact FPS but it feels fluid enough. I'd estimate around 20 fps.
